modules.md

```diff
@@ -0,0 +1,233 @@
+
+All modules within a project are nameable, there is no need for `export` anywhere on a module. If all items in a module are private (i.e., none are `export`ed), the module itself can still be named, but it is useless (like an empty object) and a glob import would import nothing.
+
+## Imports from the standard library
+
+Imports from the standard library work like other imports, but the import path begins with `std` and has no `.kcl` extension. The interpreter knows where to find the standard library which works like any other KCL project.
+
+There is an implicit import `import * from 'std'` in every file unless it includes a `@no_std` attribute. Items in `std` can be imported too. There is no explicit prelude, all of std is treated like a prelude.
+
+
+## Importing from other projects
+
+This is mostly postponed for future work. We would use the same syntax and semantics as the rest of the system, however the string in the import would include some indicator that the target is outside the project and how to locate it.
+
+Example with very, very strawman syntax: `import a from "cargo://some-library/foo.kcl"` would import `a` from the file `foo.kcl` in the project `some-library` located in the `cargo` (lol) repository. `import a from "local://../some-library/foo.kcl` would import `a` from the file `foo.kcl` from a sibling directory of the current project called `some-library`. The point of these examples is not the syntax for identifying other projects or the places where an external project might live, just to show how the `import` statement might be extended to support other projects.
+
+Privacy across multiple projects is postponed, but I would like to keep it simple, e.g., just keeping the single `export` keyword to expose names to other projects as well as other modules in the same project.
+
+### Versioning
+
+Versioning info and other extensions should all be specified within the file identifier string. Details of the design postponed.
+
+### Cycles between projects
+
+Design postponed.
+
+
+## Non-top-level imports
+
+Imports may appear anywhere within a file (including within conditionals, etc). A name may only be used following its import and at the same or narrower scope.
+
+E.g.,
+
+```
+// At top level
+import "foo.kcl"
+a = foo // ok
+
+b = bar // error
+import "bar.kcl"
+
+x = ... {
+  import "baz.kcl"
+  c = baz // ok
+  d = bar // ok
+}
+
+e = baz // error
+```
+
+Caveat: postponed design work around side-effects.
+
+
+## Importing data
+
+`import` can also be used to import data, e.g., `import "foo.json"` would create a `foo` object with the json data. Exactly how that works is postponed design, the interesting questions I see are how arbitrary data is represented in our quite limited object data structure, and how we specify what format to treat data as (is it just implied by the extension or can it be overridden? Can we support arbitrary data formats by allowing the user to provide a decoder, etc.)
```

Comment:
I really like this, great idea. I know Jordan would love that for his QR code data.